modules/schema/src/main/java/org/apache/ignite/internal/schema/CatalogSchemaManager.java

```diff
@@ -0,0 +1,487 @@
+     *
+     * @param registries Map of schema registries.
+     * @param tableId Table id.
+     * @param schema Schema descriptor.
+     * @return Future that, when complete, will resolve into an updated map of schema registries
+     *     (to be used in {@link IncrementalVersionedValue#update}).
+     */
+    private CompletableFuture<Map<Integer, SchemaRegistryImpl>> registerSchema(
+            Map<Integer, SchemaRegistryImpl> registries,
+            int tableId,
+            SchemaDescriptor schema
+    ) {
+        ByteArray key = schemaWithVerHistKey(tableId, schema.version());
+
+        byte[] serializedSchema = SchemaSerializerImpl.INSTANCE.serialize(schema);
+
+        return metastorageMgr.invoke(notExists(key), put(key, serializedSchema), noop())
+                .thenApply(t -> {
+                    SchemaRegistryImpl reg = registries.get(tableId);
+
+                    if (reg == null) {
+                        Map<Integer, SchemaRegistryImpl> copy = new HashMap<>(registries);
+
+                        copy.put(tableId, createSchemaRegistry(tableId, schema));
+
+                        return copy;
+                    } else {
+                        reg.onSchemaRegistered(schema);
+
+                        return registries;
+                    }
+                });
+    }
+
+    /**
+     * Create schema registry for the table.
+     *
+     * @param tableId Table id.
+     * @param initialSchema Initial schema for the registry.
+     * @return Schema registry.
+     */
+    private SchemaRegistryImpl createSchemaRegistry(int tableId, SchemaDescriptor initialSchema) {
+        return new SchemaRegistryImpl(
+                ver -> inBusyLock(busyLock, () -> tableSchema(tableId, ver)),
+                () -> inBusyLock(busyLock, () -> latestSchemaVersion(tableId)),
+                initialSchema
+        );
+    }
+
+    /**
+     * Return table schema of certain version from history.
+     *
+     * @param tblId Table id.
+     * @param schemaVer Schema version.
+     * @return Schema descriptor.
+     */
+    private CompletableFuture<SchemaDescriptor> tableSchema(int tblId, int schemaVer) {
+        CompletableFuture<SchemaDescriptor> fut = new CompletableFuture<>();
+
+        SchemaRegistry registry = registriesVv.latest().get(tblId);
+
+        if (registry.lastSchemaVersion() > schemaVer) {
+            return schemaByVersion(tblId, schemaVer);
+        }
+
+        CompletionListener<Map<Integer, SchemaRegistryImpl>> schemaListener = (token, regs, e) -> {
+            if (schemaVer <= regs.get(tblId).lastSchemaVersion()) {
+                SchemaRegistry registry0 = registriesVv.latest().get(tblId);
+
+                SchemaDescriptor desc = registry0.schemaCached(schemaVer);
+
+                assert desc != null : "Unexpected empty schema description.";
+
+                fut.complete(desc);
+            }
+        };
+
+        registriesVv.whenComplete(schemaListener);
```

Comment:
The current method is called only from inside the SchemaRegistryImpl, and can be moved into the registry class. This listener can be replaced with VersionTracker to wait for the next versions (like CatalogService.versionTracker).

Thus, code will be much simple. RegistersVv calls will gone as registry is already available.
Joggling with subscriptions will be replaced with versionTracker future.

Reply:
Fixed in https://issues.apache.org/jira/browse/IGNITE-20071